### DISCUSSION TITLE
ABI/API Breaking Change Detection Tools

### DISCUSSION THREAD

**Discussion by bitonality** [UE4SS-RE/RE-UE4SS#520] [General]
Regardless of what branching/release strategy we decide to use, I think we agree that we need some sort of automatic detection if ABI/API breaks will occur as a result of a PR being merged.

# What kind of changes break the ABI?

[This article](https://community.kde.org/Policies/Binary_Compatibility_Issues_With_C++) has a fairly comprehensive list of the do's and don'ts of binary compatibility. I'll include an inline snippet to give us a general sense.

You cannot...

    For existing classes:
        Unexport or remove an exported class.
        Change the class hierachy in any way (add, remove, or reorder base classes).
        Remove finality
    For template classes:
        Change the template arguments in any way (add, remove or reorder).
    For existing functions of any type:
        Unexport it.
        Remove it.
            Remove the implementation of existing declared functions. The symbol comes from the implementation of the function, so this is effectively the function.
        Inline it (this includes moving a member function's body to the class definition, even without the inline keyword).
        Add an overload (binary compatible, but not source compatible: it makes &func ambiguous). Adding overloads to already overloaded functions is ok (since any use of &func already needed a cast).

We should probably end up on the same page about what degree of ABI breaks we want our tool to detect as well as provide documentation to developers about what kind of changes they can expect to break the ABI.

# Existing Tooling

## [ABI Compliance Checker](https://lvc.github.io/abi-compliance-checker/)
Requires ELF binaries if we don't want to compile the headers with an ancient version of GCC for analysis. I played around trying to get it working purely on windows, but ran into mingw/gcc issues... Maybe someone could make more headway on this?

I really like the reports it generates. Example report for [netapi32.dll](https://abi-laboratory.pro/compatibility/Windows_10_1511_10586.494_to_Windows_10_1607_14393.0/x86_64/compat_reports/netapi32.dll/abi_compat_report.html)

## [Libabigail](https://man7.org/linux/man-pages/man7/libabigail.7.html)
Requires ELF binaries
[Simple example](https://developers.redhat.com/blog/2020/04/02/how-to-write-an-abi-compliance-checker-using-libabigail#the_example_project__libslicksoft_so)
```bash
$ make abi-check
../tools/check-abi ../abi-ref/libslicksoft.so.abi libslicksoft.so || echo "ABI compatibility issue detected!"
An incompatible ABI change was detected:

Functions changes summary: 1 Removed, 0 Changed, 0 Added function
Variables changes summary: 0 Removed, 0 Changed, 0 Added variable

1 Removed function:

  'function void function_1()'    {function_1}

ABI compatibility issue detected!
$
```

# Custom Tooling

## Dumpbin

We could leverage `dumpbin /exports` to compare API level symbols. This is not full ABI compatibility, but it's a start that could work for us pretty much out of the box.

# Other tools

I'll update this original post as people make more suggestions/considerations.

**Comment by Buckminsterfullerene02** [comment]
Nice post, this will be useful :D

So it's more of a compatibility scale rather than it is or it isn't. So for C++ mods, if the ABI changes on the UE4SS.dll at all, they would break, or only if they are trying to touch certain parts of it? It seems to me like it would be a binary classification for this scenario. 

For "other tools", there aren't any more I know of, *however*, in the worst case scenario where none of the above work out, I'm pretty sure the only option involves building a C++ mod and UE4SS and then installing both onto a blank shipping game on a VM and checking the logs to see if the C++ mod gets the 0x7f error that means it failed to load due to ABI incompatibility.

**Reply by UE4SS** [reply]
Didn't someone at some point a while back look into running a headless UE instance in a github action ?
We could presumably still do that with our own hosted runner if the builtin ones are too tightly locked down.

Would using our own hosted runner help with the PR/CI problem as well ? I don't know how much freedom a self hosted runner has compared to the ones provide by github but if we can set it up to use its own set of keys and executing the checks in its own repo instead of trying to use the actual PR branch we'd sidestep this problem.
If not, we'd need to have a VM or something and just use the action as a trigger for the VM to sidestep the problem.
Otherwise we have to fix this problem somehow.

**Comment by bitonality** [comment]
Currently working on the ABI breakage detector in rust by using https://crates.io/crates/pdb
It uses the PDB to generate an intermediate snapshot of the ABI at a point in time. We can then feed two of these snapshots to the tool in order to diff and detect ABI breakage. 

I'm fond of the idea of generating an intermediate file since the comparison tool will only have to download two of the intermediate files instead of two PDBs. It also let us run the build wherever we can (the pseudo licensing requirement), generate the ABI intermediates from that same machine, and then upload the intermediates as an artifact. We can then just download artifacts for the *new* and *old* versions of the API and run the rust tool to detect breakage against those artifacts.

I've wired up all the following cases that should flag ABI breakage
```rust
pub enum ClassProblem {
    RemovedOrUnexported,
    HierarchyChanged,
    FinalityRemoved,
    FuncProblem(FuncSigProblem),
    FieldProblem(FieldSigProblem),
    TemplateProblem(TemplateProblem),
}

pub enum FuncSigProblem {
    RemovedOrUnexported,
    InlineAdded,
    OverloadIntroduced,
    CvQualifierChanged,
    AccessRightsChanged,
    ReturnTypeChanged,
    VirtualProblem(VirtualMemberFuncProblem)
}

pub enum VirtualMemberFuncProblem {
    Introduced,
    Added,
    Removed,
    OverrideNonPrimaryBase,
    OverrideCovariantReturnType,
}

pub enum FieldSigProblem {
    Added,
    Removed,
    TypeChanged,
    StaticSigFieldProblem(StaticFieldSigProblem)
}

pub enum StaticFieldSigProblem {
    RemovedOrUnexported,
    TypeChanged,
    CvQualifierChanged,
}

pub enum TemplateProblem {
    ArgumentsChanged
}
```

**Reply by UE4SS** [reply]
How is the intermediate file for a PR supposed to be generated ?
Would we set up a separate server with the proper key that pulls and builds ?
Do we merge blindly and do the comparison at that point, hoping that nothing is broken ?
We need to figure this out because we can't run any build action directly on a PR right now.

**Reply by bitonality** [reply]
Why don't we just store the SSH info as a secret within the organization? 
https://docs.github.com/en/actions/security-guides/using-secrets-in-github-actions?tool=webui

Then we can just set up our runners to consume the secret as part of the workflow.
```
steps:
  - name: Hello world action
    with: # Set the secret as an input
      super_secret: ${{ secrets.SuperSecret }}
    env: # Or as an environment variable
      super_secret: ${{ secrets.SuperSecret }}
```

Secrets do not get passed to runners that are triggered from a fork, BUT this shouldn't matter to us since PRs from forked Repo->UE4SS trigger runners from UE4SS. As long as the PR is going INTO UE4SS, then the runners are triggered from UE4SS and not from the forks.

Basic flow for validating whether a PR should be merged would be:
1. PR Build Queued (as a pre-merge check)
2. Github waits for the next available runner from the runner pool.
3. A runner picks up the job
4. Runner automatically retrieves the secret from the organization's secret store.
5. Runner does the merge of UE4SS/UEPseudo from fork branch -> target branch
6. Runner executes a full UE4SS build
7. If build fails, indicate to the workflow that PR completion is blocked until the build succeeds. Exit.
8. If build succeeds, proceed to ABI checking.
9. If the PR has been flagged by a maintainer that breaking changes are okay, indicate to the workflow that PR completion is unblocked. Exit.
10. Download the latest ABI artifact from target branch (essentially the ABI artifact that was generated for the most recent commit to the target branch).
11. Run the ABI tool comparing the downloaded artifact to the PDB we generated from our build.
12. If no breaking changes detected, indicate to the workflow that PR completion is unblocked. Exit.
13. If breaking changes detected, indicate to the workflow that PR completion is blocked unless a maintainer indicates in the PR that the breaking change is acceptable. Exit.

If the above checks pass and PR completion is unblocked, then the merge workflow would look like:
1. PR Merge Queued
2. Github waits for the next available runner from the runner pool.
3. A runner picks up the job
4. Runner automatically retrieves the secret from the organization's secret store.
5. Runner does the merge of UE4SS/UEPseudo from fork branch -> target branch
6. Runner executes a full UE4SS build
7. If build fails, indicate potential transient failure (since it already passed validation). Exit.
8. If build succeeds run the ABI tool on the build output.
9. Upload the ABI tool result as an artifact (this is what will be compared against for the next PR that someone opens).
10. Complete the merge and close the PR.

I don't think we'll need any custom tooling other than the ABI tool. All the steps I've outlined have existing tools/workflows baked into the GitHub CI stack natively. 

Do you see any glaring issues with this proposed solution?

**Reply by UE4SS** [reply]
> As long as the PR is going INTO UE4SS, then the runners are triggered from UE4SS and not from the forks.

Are you absolutely sure about this ?
If this was the case, I don't think we would be having a problem right now as it would  be able to use the `secrets.UEPSEUDO_PAT` key that we already have or am I wrong here ?
It runs it on the fork which doesn't have `secrets.UEPSEUDO_PAT` and that's why it fails as far as I understand.
I'm very sure we've had errors because of that before, but I suppose one way to make sure is to just enable the experimental release action on PRs (because it seems you can't manually trigger on fork PRs) and see if it works.

**Reply by bitonality** [reply]
> > As long as the PR is going INTO UE4SS, then the runners are triggered from UE4SS and not from the forks.
> 
> Are you absolutely sure about this ? If this was the case, I don't think we would be having a problem right now as it would be able to use the `secrets.UEPSEUDO_PAT` key that we already have or am I wrong here ? It runs it on the fork which doesn't have `secrets.UEPSEUDO_PAT` and that's why it fails as far as I understand. I'm very sure we've had errors because of that before, but I suppose one way to make sure is to just enable the experimental release action on PRs (because it seems you can't manually trigger on fork PRs) and see if it works.

For pull requests from a forked repository to the base repository, GitHub sends the pull_request, issue_comment, pull_request_review_comment, pull_request_review, and pull_request_target events to the base repository. No pull request events occur on the forked repository.

Source: https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull-request-events-for-forked-repositories

**Reply by UE4SS** [reply]
Then we should just enable the experimental builds for PRs to test it.
@localcc and @narknon are the people with the ability to do that.

**Reply by UE4SS** [reply]
Actually, is this just a workflow file change ? I have forgotten how this works.

**Reply by narknon** [reply]
How does this work if someone PRs to uepseudo from their own fork?

**Reply by UE4SS** [reply]
According to @bitonality nothing runs on the fork, it runs on UE4SS/UEPseudo.
I want to test it so that we know for sure because I have my doubts, but I hope those are wrong.

**Reply by narknon** [reply]
But the fork is where the changes are, does it create a branch in the main repo with the changes from the PR merged in?

**Reply by UE4SS** [reply]
No, that's why I doubt it will work.

**Reply by bitonality** [reply]
It might benefit this discussion to look into the process that git uses under the hood during pull requests if you are unfamiliar.

In short, each PR has its own remote ref that's created by github in the form of `refs/pull/239/merge` and `refs/pull/239/head`. These are in the base repo (UE4SS repo in this case).

This is how workflows/runners are able to run against the correct PR context when they are triggered.
